### ISSUE TITLE
pp=2 bit-exactness gate has been red on gfx1201 since 2026-06-02 (DeltaNet Q8 dither frame counter is topology-dependent)

### ISSUE THREAD

**Issue by Kaden-Schutt** [Kaden-Schutt/hipfire#588]
`crates/hipfire-arch-qwen35/tests/pp_parity.rs` — the pp=2 bit-exactness gate whose own docstring reads *"This is the floor — if this regresses, pp=2 is broken"* — has been **failing on gfx1201 since 2026-06-02** and nobody noticed, because it is `#[ignore]`d and gated behind `HIPFIRE_HAVE_2_GPU=1`, so CI never runs it.

Found while validating the multi-GPU blockers that #527 asks contributors for.

## Reproduction

4x R9700 (gfx1201), ROCm 7.14, `origin/master` @ `687c6509`, fixture `qwen3.5-0.8b.mq4`:

```
HIP_VISIBLE_DEVICES=0,1 HIPFIRE_HAVE_2_GPU=1 \
  cargo test -p hipfire-arch-qwen35 --release --features deltanet \
             --test pp_parity -- --ignored
```

```
pp parity broke at decode step 3: pp=1=717, pp=2=8340
pp=1 head: [248068, 198, 90700]
pp=2 head: [248068, 198, 90700]
```

Deterministic — byte-identical across repeated runs.

## Bisect

`git bisect` over the 2280 commits between the test's introduction and 2026-07-28, run automatically against the real hardware:

- **PASS** at `029a365f` (2026-05-05, alpineq, *"pp examples + pp-gate + pre-commit hook — Stage 8 of #58"*)
- **FAIL** at `fdf1d46c` (2026-07-28)
- **First bad commit: `2159e05103020918ca109de21db9093220c51494`** (2026-06-02, @nwoolmer) — *"fix(qwen3.6-a3b): serving robustness + DeltaNet Q8 dither + OpenAI penalties"*

Note this is **not** a #549 / FireMap regression — `94d0bf2b6^` already fails.

## Root cause

That commit reseeded the Q8 GatedDeltaNet stochastic-rounding dither. From `kernels/src/gated_delta_net_q8.hip:24`:

```c
unsigned int frame,  // per-launch monotonic counter (host-supplied)
...
// Seed = spatial location (h, row, tid) + per-launch `frame` counter.
+ (frame + (unsigned int)bt) * 2531011u);
```

`frame` comes from a **process-global** `AtomicU32` (`crates/rdna-compute/src/norm.rs:20`) via `reserve_gdn_requant_frames`. The seed is therefore a function of *launch order*, which is topology-dependent.

Two distinct problems fall out.

### 1. The test never aligns the two arms' frame streams

`run_pp1` and `run_pp2` execute in the same process, so pp=2 starts mid-stream wherever pp=1 left the counter. The codebase already has the API for exactly this — `gdn_requant_frame_checkpoint()` / `restore_gdn_requant_frame_checkpoint()` (`norm.rs:30-38`, *"so two mutually exclusive arms consume identical stochastic-rounding frame IDs"*), used in ~10 places in `examples/daemon.rs`. The test does not use it.

Adding it is **necessary but not sufficient** — patch at the end of this issue. Proof it is necessary and that the engine is otherwise deterministic: with the checkpoint in place, swapping the second arm to a second `run_pp1` call **passes**.

```
=== pp1 vs pp1, same process, frame restored ===
test result: ok. 1 passed; 0 failed
```

### 2. The multi-GPU path reserves 32x the frames (the actual bug)

With the comparison made sound, instrumenting frame consumption per arm:

```
GDN frames consumed: pp1=36  pp2=1152
```

**Exactly 32x.** The pp=2 path reserves stochastic-rounding frames on a completely different schedule than pp=1 — this reads like per-element/per-head reservation where the single-GPU path reserves per-launch. Because the seed mixes `frame` in directly, differing reservation granularity guarantees a different dither stream on every requant, so **bit-exact pp1 == pp2 is structurally unattainable** regardless of how correct the rest of the pp=2 path is.

With the frame streams aligned at the start, divergence actually moves *earlier* (step 1, `pp=1=198` vs `pp=2=271`) because the 32x drift begins immediately rather than after the arms happen to resynchronise.

## Why this matters now

#527 lists *"Qwen35 arch-resident PP needs the same physical residency and transfer validation"* as a merge blocker. That work will be layered on a PP path whose bit-exactness gate is red and has been for ~3 months. This should be resolved before, or as part of, landing that PR.

## Suggested direction

Seed the dither from a **topology-invariant** quantity — global layer index + token position, alongside the existing `(h, row, tid)` — instead of a launch-order counter. That preserves the property the original commit was protecting (dither independent of the value being rounded, fresh per requant) while making it reproducible across pp=1/pp=2, replay, and device-count changes. The 32x reservation asymmetry should be investigated on its own regardless, since it is unexplained.

## Also worth fixing: the gate cannot fail visibly

`#[ignore]` + env gate means a red floor stays silent indefinitely. Options: a nightly multi-GPU CI job, or wiring it into `scripts/pp-gate.sh` (referenced in the test's own comments as *"the canonical hands-free regression"*) with a runner that has 2 GPUs.

## Patch for problem 1 (necessary, not sufficient)

```diff
+    // The Q8 GatedDeltaNet stochastic-rounding dither is seeded from a
+    // process-global monotonic frame counter (`GDN_REQUANT_FRAME`), not from
+    // the value being rounded. Both arms therefore have to consume *identical*
+    // frame IDs, or their requant dither diverges and bit-exact parity is
+    // unsatisfiable no matter how correct the pp=2 path is. `run_pp1` runs
+    // first and advances the counter, so pp=2 would otherwise start mid-stream.
+    // The daemon's own A/B arms take this same checkpoint/restore (see the
+    // `gdn_requant_frame_checkpoint` callsites in `examples/daemon.rs`).
+    let frame_checkpoint = rdna_compute::norm::gdn_requant_frame_checkpoint();
     let toks_pp1 = run_pp1(&model, &prompt);
+    rdna_compute::norm::restore_gdn_requant_frame_checkpoint(frame_checkpoint);
     let toks_pp2 = run_pp2(&model, &prompt);
```

Not opening this as a PR since it leaves the gate red; it belongs with the real fix.

cc @nwoolmer (dither seeding), @alpineq (original pp gate)

**Comment by Kaden-Schutt** [comment]
Follow-up measurement that **partly retracts the root-cause attribution above**. Recording it here and pausing the investigation.

## What I ran

Same fixture and hardware as the original report, varying only graph capture, with per-arm frame consumption instrumented:

| Config | pp1 frames | pp2 frames | Divergence |
|---|---:|---:|---|
| default | 36 | 1152 | step 1, `pp=1=198` `pp=2=271` |
| `HIPFIRE_VERIFY_GRAPH=0` | 36 | 1152 | step 1, `pp=1=198` `pp=2=271` |
| `HIPFIRE_GRAPH=0 HIPFIRE_VERIFY_GRAPH=0` | **1152** | 1152 | step 1, `pp=1=198` `pp=2=271` |

## Correction: frame accounting does not explain the divergence

With graph capture off, pp1 and pp2 consume **identical** frame counts (1152 = 18 LinearAttention layers x 64 forward passes, which is the expected one-per-launch rate). The divergence is **bit-identical** in that configuration — same step, same two token IDs.

Equal frame streams, unchanged failure. So the 32x asymmetry I flagged as "the actual bug" is **not** what drives the pp1/pp2 token divergence. My analysis above over-attributed it, and the "suggested direction" (topology-invariant seeding) does not follow from this evidence. Treat that section as withdrawn pending a real mechanism.

Corollary: pp1's emitted token at step 1 is `198` whether it consumed 36 frames or 1152. The dither frame is not moving pp1's argmax at all on this fixture, which is the cleanest single piece of evidence against the dither theory.

## What does survive

**1. Graph capture freezes the GDN stochastic-rounding frame.** pp1 reserves 36 frames (18 layers x 2) across a 64-pass run under capture, versus 1152 uncaptured. So under a captured graph the dither scalar is baked at capture time and reused for every replay, which is the opposite of the "fresh per requant" property `2159e051` was written to provide. Notably `crates/rdna-compute/src/replay.rs:2651` guards exactly this hazard for the direct-AQL path — *"so bypassing the Rust launch wrapper does not freeze the captured frame scalar"* — so the hazard is known; the hipGraph path appears not to have the same guard. Whether that is a defect or accepted behaviour needs an owner's call. It is independent of pp parity and would apply to ordinary single-GPU captured decode.

**2. The bisect still stands, the mechanism does not.** `2159e051` is reproducibly the first bad commit (`029a365f` passes, `94d0bf2b6^` fails). But since dither-frame accounting is now ruled out as the driver, the responsible change is something *else* in that commit — it also touched `crates/rdna-compute/src/norm.rs`, `sampling.rs`, `sampler.rs`, `kernels/src/sample_top_p.hip`, and non-dither lines of `kernels/src/gated_delta_net_q8.hip`. The test argmaxes downloaded logits and never calls the sampler, so the sampling-side changes should be inert here — that narrows it to the `norm.rs` / GDN-kernel side, unverified.

**3. The test still needs the checkpoint/restore.** Independently justified: with it, swapping the second arm to a second `run_pp1` **passes**, which establishes the harness is otherwise deterministic and makes the comparison sound. It is a prerequisite for any further bisect *within* `2159e051`, not a fix.

## Status

Pausing here rather than guessing further. Next step for whoever picks this up is a within-commit bisect of `2159e051` (revert the GDN kernel hunks and the `norm.rs` hunks separately against current beta) to find which change actually moves pp=2. The `HIPFIRE_GRAPH=0` configuration is the right one to bisect under, since it removes the frame-count confound entirely.

The gate remains red on gfx1201 and remains invisible to CI (`#[ignore]` + `HIPFIRE_HAVE_2_GPU`), which is the part worth fixing regardless of mechanism.
